src/miniscript/musig_key.rs

```diff
@@ -0,0 +1,166 @@
+//! Support for multi-signature keys
+use core::fmt;
+use core::str::FromStr;
+
+use crate::expression::{FromTree, Tree};
+use crate::prelude::*;
+use crate::{Error, MiniscriptKey};
+
+#[derive(Clone, PartialEq, PartialOrd, Ord, Eq, Hash)]
+/// Enum for representing musig keys in miniscript
+pub enum KeyExpr<Pk: MiniscriptKey> {
+    /// Single-key (e.g pk(a), here 'a' is a single key)
+    SingleKey(Pk),
+
+    /// Collection of keys in used for musig-signature
+    MuSig(Vec<KeyExpr<Pk>>),
+}
+
+impl<Pk: MiniscriptKey + FromStr> FromTree for KeyExpr<Pk> {
+    fn from_tree(tree: &Tree) -> Result<KeyExpr<Pk>, Error> {
+        if tree.name == "musig" {
+            let mut key_expr_vec = vec![];
+            for sub_tree in tree.args.iter() {
+                let temp_res = KeyExpr::<Pk>::from_tree(sub_tree)?;
+                key_expr_vec.push(temp_res);
+            }
+            Ok(KeyExpr::MuSig(key_expr_vec))
+        } else {
+            let single_key = Pk::from_str(tree.name).map_err(|_| Error::SingleKeyParseError)?;
+            Ok(KeyExpr::SingleKey(single_key))
+        }
+    }
+}
+
+impl<Pk: MiniscriptKey + FromStr> FromStr for KeyExpr<Pk> {
+    type Err = Error;
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let (key_tree, _) = Tree::from_slice(s).unwrap();
+        FromTree::from_tree(&key_tree)
+    }
+}
+
+#[derive(Debug, Clone)]
+/// Iterator for keyexpr
+pub struct KeyExprIter<'a, Pk: MiniscriptKey> {
+    stack: Vec<&'a KeyExpr<Pk>>,
+}
+
+impl<'a, Pk> Iterator for KeyExprIter<'a, Pk>
+where
+    Pk: MiniscriptKey + 'a,
+{
+    type Item = &'a Pk;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        while !self.stack.is_empty() {
+            let last = self.stack.pop().expect("Size checked above");
+            match &*last {
```

Comment:
nit: Since we don't have any overloaded `Deref` for `KeyExpr`, ig it's safe to just match it as:
```suggestion
            match last {
```

Reply:
Ok, changed accordingly.
```suggestion
            match last {
```